File: homing-relation-grid/src/main/java/hue/captains/singapura/js/homing/grid/GridUpdateBatchModule.java
package hue.captains.singapura.js.homing.grid;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * RFC 0050 — {@code GridUpdateBatch}: the direct update path's queue, split
 * out of the facade by the line ratchet. One concern — coalesce domain-pushed
 * cell updates per animation frame, last write per cell wins, so a hot feed
 * (a popularity tick, a Minesweeper flood fill) collapses to one
 * {@code cell.update()} per cell per frame. Falls back to a synchronous flush
 * where no {@code requestAnimationFrame} exists (headless). {@code drain()}
 * is the read-path guard: a hidden page gets no frames, so anything that
 * READS cell state (copy, export) must drain first or read stale values.
 * No DOM, no maps, no layout.
 */
public record GridUpdateBatchModule() implements DomModule<GridUpdateBatchModule> {

    /** The single export — the {@code GridUpdateBatch} JS class. */
    public record GridUpdateBatch() implements Exportable._Constant<GridUpdateBatchModule> {}

    public static final GridUpdateBatchModule INSTANCE = new GridUpdateBatchModule();

    @Override
    public ImportsFor<GridUpdateBatchModule> imports() {
        return ImportsFor.noImports();
    }

    @Override
    public ExportsOf<GridUpdateBatchModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new GridUpdateBatch()));
    }
}

File: homing-relation-grid/src/main/resources/homing/js/hue/captains/singapura/js/homing/grid/GridUpdateBatchModule.js
// =============================================================================
// GridUpdateBatchModule — the DIRECT update path's queue, split out of the
// facade when the line ratchet forced it (third time; it keeps finding real
// seams). One concern: coalesce domain-pushed cell updates per animation
// frame, LAST WRITE PER CELL WINS, so a hot feed (a popularity tick, a
// Minesweeper flood fill) collapses to one cell.update() per cell per frame.
//
//   · No requestAnimationFrame (headless, or a hidden page at boot) → the
//     flush runs synchronously, so tests and SSR-ish contexts behave.
//   · drain() is the READ-PATH guard: a hidden page gets no frames, so
//     pending writes can sit indefinitely — harmless for painting, fatal for
//     anything that reads cell state (copy, export). Read paths drain first.
//
// No DOM, no maps, no layout: it holds a Map and calls cells.update.
// =============================================================================

class GridUpdateBatch {

    constructor(deps) {
        deps = deps || {};
        if (!deps.cells) throw new Error("[GridUpdateBatch] cells is required");
        this._cells = deps.cells;
        this._pending = new Map();      // key → { pk, col, value }
        this._scheduled = false;
    }

    /** Queue a value for (pk, col); unknown cells are ignored (not an error —
     *  a filtered-out row still has cells, a deleted one does not). */
    push(pk, col, value) {
        if (!this._cells.get(pk, col)) return false;
        this._pending.set(pk + " " + col, { pk: pk, col: col, value: value });
        this._schedule();
        return true;
    }

    _schedule() {
        if (this._scheduled) return;
        this._scheduled = true;
        var self = this;
        var raf = (typeof requestAnimationFrame === "function") ? requestAnimationFrame
                : function (fn) { fn(); };
        raf(function () { self._scheduled = false; self.drain(); });
    }

    /** Apply everything queued, now. Safe to call at any time. */
    drain() {
        var self = this, pending = this._pending;
        this._pending = new Map();
        pending.forEach(function (u) { self._cells.update(u.pk, u.col, u.value); });
        return this;
    }

    pendingCount() { return this._pending.size; }
}
